File: algorithm/pascals-triangle.cpp
118. Pascal Triangle
Difficulty: Easy
Given numRows, generate the first numRows of Pascal triangle.

For example, given numRows = 5,
Return

[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]

// Time:  O(n^2)
// Space: O(1)

class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> result;
        for (int i = 0; i < numRows; ++i) {
            result.push_back({});
            for (int j = 0; j <= i; ++j) {
                if (j == 0 || j == i) {
                    result[i].emplace_back(1);
                } else {
                    result[i].emplace_back(result[i - 1][j - 1] +
                                           result[i - 1][j]);
                }
            }
        }
        return result;
    }
};


class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> result;
        vector<int> array;
        for (int i = 0; i < numRows; i++) {
            for (int j = i - 1; j > 0; j--) {
                array[j] = array[j - 1] + array[j];
            }
            array.push_back(1);
            result.push_back(array);
        }
        return result;
    }
}

File: zZ/java/pascals-triangle.java
/**
 * Pascal's Triangle
 * 
 * Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.
 * 
 * 
 * In Pascal's triangle, each number is the sum of the two numbers directly above it.
 * 
 * Example:
 * 
 * Input: 5
 * Output:
 * [
 *      [1],
 *     [1,1],
 *    [1,2,1],
 *   [1,3,3,1],
 *  [1,4,6,4,1]
 * ]
 */

import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> result = new ArrayList<List<Intger>>();
        List<Integer> entry = new ArrayList<Integer>();
        for (int i = 0; i < numRows; ++i) {
            for (int j = i - 1; j > 0; --j) {
                entry.set(j, entry.get(j) + entry.get(j - 1));
            }
            entry.add(1);
            result.add(new ArrayList<Integer>(entry));
        }
        return result;
    }
}
